### ISSUE TITLE
"KeyError: <class 'sqlalchemy.types.Interval'>" after server disconnected from Oracle 8 database

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2411]
_Imported from trac ticket 2411.
Created by **mariam**
Opened in trac: Jul 12, 2013, 8:23:36 AM
Last modified in trac: Jul 12, 2013, 9:23:20 AM_
***
_by mariam:_
See http://www.sqlalchemy.org/trac/ticket/2776

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 12, 2013, 9:23:20 AM:_

(Changeset [[4335]](../commit/178c9b85c177f734a97ad86423a99da117cfaaf9))
* fixes #2411, "KeyError: <class 'sqlalchemy.types.Interval'>" after server disconnected from Oracle 8 database #eld #isf

 * patches fix for ticket http://www.sqlalchemy.org/trac/ticket/2776, see #1550
```diff
diff -U10 -r SQLAlchemy-0.7.604kb/lib/sqlalchemy/engine/strategies.py SQLAlchemy-0.7.605kb/lib/sqlalchemy/engine/strategies.py
--- SQLAlchemy-0.7.604kb/lib/sqlalchemy/engine/strategies.py    2012-03-21 15:25:46.000000000 -0400
+++ SQLAlchemy-0.7.605kb/lib/sqlalchemy/engine/strategies.py    2013-07-10 01:52:14.000000000 -0400
@@ -148,20 +148,21 @@
             if do_on_connect:
                 def on_connect(dbapi_connection, connection_record):
                     conn = getattr(dbapi_connection, '_sqla_unwrap', dbapi_connection)
                     if conn is None:
                         return
                     do_on_connect(conn)

                 event.listen(pool, 'first_connect', on_connect)
                 event.listen(pool, 'connect', on_connect)

+            @util.only_once
             def first_connect(dbapi_connection, connection_record):
                 c = base.Connection(engine, connection=dbapi_connection)

                 # TODO: removing this allows the on connect activities
                 # to generate events.  tests currently assume these aren't
                 # sent.  do we want users to get all the initial connect
                 # activities as events ?
                 c._has_events = False

                 dialect.initialize(c)
diff -U10 -r SQLAlchemy-0.7.604kb/lib/sqlalchemy/__init__.py SQLAlchemy-0.7.605kb/lib/sqlalchemy/__init__.py
--- SQLAlchemy-0.7.604kb/lib/sqlalchemy/__init__.py     2012-06-06 18:03:49.000000000 -0400
+++ SQLAlchemy-0.7.605kb/lib/sqlalchemy/__init__.py     2013-07-10 01:41:31.000000000 -0400
@@ -110,16 +110,16 @@
     ThreadLocalMetaData,
     UniqueConstraint,
     )

 from sqlalchemy.engine import create_engine, engine_from_config


 __all__ = sorted(name for name, obj in locals().items()
                  if not (name.startswith('_') or inspect.ismodule(obj)))

-__version__ = '0.7.604kb'
+__version__ = '0.7.605kb'

 del inspect, sys

 from sqlalchemy import util as _sa_util
 _sa_util.importlater.resolve_all()
diff -U10 -r SQLAlchemy-0.7.604kb/lib/sqlalchemy/util/__init__.py SQLAlchemy-0.7.605kb/lib/sqlalchemy/util/__init__.py
--- SQLAlchemy-0.7.604kb/lib/sqlalchemy/util/__init__.py        2012-03-21 15:25:46.000000000 -0400
+++ SQLAlchemy-0.7.605kb/lib/sqlalchemy/util/__init__.py        2013-07-10 01:54:36.000000000 -0400
@@ -20,15 +20,16 @@
 from langhelpers import iterate_attributes, class_hierarchy, \
     portable_instancemethod, unbound_method_to_callable, \
     getargspec_init, format_argspec_init, format_argspec_plus, \
     get_func_kwargs, get_cls_kwargs, decorator, as_interface, \
     memoized_property, memoized_instancemethod, \
     reset_memoized, group_expirable_memoized_property, importlater, \
     monkeypatch_proxied_specials, asbool, bool_or_str, coerce_kw_type,\
     duck_type_collection, assert_arg_type, symbol, dictlike_iteritems,\
     classproperty, set_creation_order, warn_exception, warn, NoneType,\
     constructor_copy, methods_equivalent, chop_traceback, asint,\
-    generic_repr, counter
+    generic_repr, counter,\
+    only_once

 from deprecations import warn_deprecated, warn_pending_deprecation, \
     deprecated, pending_deprecation

diff -U10 -r SQLAlchemy-0.7.604kb/lib/sqlalchemy/util/langhelpers.py SQLAlchemy-0.7.605kb/lib/sqlalchemy/util/langhelpers.py
--- SQLAlchemy-0.7.604kb/lib/sqlalchemy/util/langhelpers.py     2012-03-21 15:25:46.000000000 -0400
+++ SQLAlchemy-0.7.605kb/lib/sqlalchemy/util/langhelpers.py     2013-07-10 01:58:09.000000000 -0400
@@ -866,20 +866,34 @@
        runs, with a compatible version that uses
        warnings.warn_explicit, so that the warnings registry can
        be controlled.

     """
     if isinstance(msg, basestring):
         warnings.warn(msg, exc.SAWarning, stacklevel=stacklevel)
     else:
         warnings.warn(msg, stacklevel=stacklevel)

+
+def only_once(fn):
+    """Decorate the given function to be a no-op after it is called exactly
+    once."""
+
+    once = [fn]
+    def go(*arg, **kw):
+        if once:
+            once_fn = once.pop()
+            return once_fn(*arg, **kw)
+
+    return update_wrapper(go, fn)
+
+
 _SQLA_RE = re.compile(r'sqlalchemy/([a-z_]+/){0,2}[a-z_]+\.py')
 _UNITTEST_RE = re.compile(r'unit(?:2|test2?/)')
 def chop_traceback(tb, exclude_prefix=_UNITTEST_RE, exclude_suffix=_SQLA_RE):
     """Chop extraneous lines off beginning and end of a traceback.

     :param tb:
       a list of traceback lines as returned by ``traceback.format_stack()``

     :param exclude_prefix:
       a regular expression object matching lines to skip at beginning of ``tb``
```

**Comment by kentbower** [comment]
_**kbower** trac status change on Jul 12, 2013, 9:23:20 AM:_ **closed**
